A few nextcloud install fixes

jbz-dev opened the pull request:
Ref: https://github.com/Fourdee/DietPi/pull/1142

---

( In Progress )

- [X] php-intl installation _( Fixed )_
- [X] Nginx configuration rechecked and working _( Working )_
- [X] HTST not set correctly _( Not nextcloud centric )_
- [X] HTTP-Splitting vulnerability _( Fixed )_
- [X] Opcache error on nextcloud admin page _( Fix available )_

 
Nextcloud installation failed because of `AGI "$php_package_version"-intl` was interpreted as an installation with the argument `-intl`.

I corrected the install process by splitting the installation

- If Stretch -> `AGI php-intl`
- If Jessie -> `AGI php5-intl`

For the nextcloud nginx configuration file I assume that most users want their instance to be available at `theirdomainname.com/nextcloud` so instead of creating a virtualhost in `sites-available` it's more useful to put the `nextcloud.conf` in _sites-dietpi_ directory.

jbz-dev committed:
Fix php-intl installation

- Installing php-intl for Stretch and php5-intl for Jessie

jbz-dev committed:
Fix Nginx nextcloud conf file location

  - mv `nextcloud.conf` from `sites-available` to `sites-dietpi`
  - rm symlink between sites-available and sites-enabled for nextcloud.conf

jbz-dev commented:
HTST should be set on the global nginx configuration as it is a setting that is not only useful for nextcloud but the whole server.

So I think it is not a nextcloud related issue.

MichaIng commented:
Thanks for fix! 🙂

- Still wondering why the AGI was not working with you. @Fourdee did you test the PR before and did it also not work for you? I did not run the whole dietpi-software, but copy&paste variable and install/activation part worked fine for me.
And in case what did I wrong, as the other PHP modules are installed the same way 😕?
But if it works the way here, it's fine. Little more code 😆.

- Yes HSTS should not be set here. We don't know if/how the user want or don't want to use SSL. It will be configured by new certbot installation automatically on nginx and apache.

- OPCache errors should disappear after nextcloud.conf is working on nginx. It contains the wanted settings then.

- Yeah installation should be in domain.com/nextcloud, as this is already default on DietPi, also for current (master) Nextcloud installation (and ownCloud btw).

jbz-dev commented:
@MichaIng 

Thanks for the answer.

I looked upon opcache values this morning and even if they are in `nextcloud.conf` with the lines added with your PR they do not change in `php.ini` and the error message is still here on Nextcloud.

jbz-dev committed:
Fix Possible HTTP-Splitting vulnerability 

- Problem: [http_splitting] Possible HTTP-Splitting vulnerability.
- Description: Using variables that can contain "\n" may lead to http injection.
- Additional info: https://github.com/yandex/gixy/blob/master/docs/en/plugins/httpsplitting.md
- Reason: At least variable "$uri" can contain "\n"

jbz-dev commented:
One thing we can do to ensure that opcache is only been set for Nextcloud is to

- Create a `nextcloud.conf` and `nextcloud.ini` in `php/fpm/conf.d`
- Append the nextcloud recommended values for opcache in `nextcloud.ini`
- Edit the NC nginx configuration file to listen to the nextcloud php-fpm socket.

I've just tested it and it's working really great. (The idea was taken from [Yunohost](https://github.com/YunoHost-Apps/nextcloud_ynh/blob/master/scripts/install)

MichaIng commented:
Sounds good. I would like to create just one file for nextcloud instead of two, to keep it simpler and everything related together. But good to have an alternative.

Could you, with php configuration in nextcloud.conf, test if non of the php values is set? You could use DietPi's `opcache.php`, `apc.php` and `phpinfo.php` after putting them into nextcloud directory. If really non of them is working, then there seems to be some syntax/method error. I was looking around, generally the solution should work. However there is another, bid simpler way to add multiple lines to one fastgci_param (PHP_ADMIN_VALUE) by manual line break ([source](https://coderwall.com/p/gt2g3q/setting-multiple-php_value-in-nginx-config)):

```
fastcgi_param PHP_ADMIN_VALUE "opcache.enable_cli=1
opcache.interned_strings_buffer=8
opcache.max_accelerated_files=10000
opcache.memory_consumption=128
opcache.save_comments=1
opcache.revalidate_freq=1
mysql.allow_local_infile=1
mysql.allow_persistent=1
mysql.max_persistent=-1
mysql.max_links=-1
mysql.default_port=0
mysql.default_socket=/var/run/mysqld/mysqld.sock
mysql.connect_timeout=60
mysql.trace_mode=0
apc.enable_cli=1";
```

Could your try to add this instead of https://github.com/Fourdee/DietPi/blob/testing/dietpi/dietpi-software#L8572-L8599?

jbz-dev commented:
Nope not working, I already tested that config this morning. I really don't know what we are missing and I can't find any further informations online about properly setting opcache on nginx.

MichaIng commented:
As opcache is a php module, it is basically the question how to add php values to nginx configuration files. `fastcgi_param PHP_ADMIN_VALUE` / `PHP_VALUE` is the way to go. I also saw examples using this inside directory directives, so the place there should be allowed.

Only thing to take care is, that each `fastcgi_param` can only be set one time, so setting it multiple times (with different php values) would just override one by another. So you have to add it as one string with multiple lines, eather separated by `\n` (my first attempt) or by manual line breaks (attempt above). For both I found examples claimed to work, one even inside php.net manual. So I have no idea why it is not working here.

One thing is, that I am not sure about the mysql values, if they work the way intended (as nextcloud uses pdo_mysql module) and if they even exist. Where apache simply ignores wrong php values there, nginx might break with them inside the string. `mysql.cache_size` was given in nextcloud admin manual but not listed on php.net, so I removed it above. One last try could be to remove all `mysql.` from the string. All other values (opcache+apc) definitely exist, if apcu and opcache modules are installed an active.

jbz-dev commented:
Little bit of testing and here are the results : 

- fastcgi_param PHP_ADMIN_VALUE not working properly where located inside a location folder (in nextcloud nginx conf for example)
- However the parameters are good when I put them inside the default nginx conf `sites-available/default" and Nexcloud security test passes without any complains.

Will digg on the subject later today if I have time to.

MichaIng commented:
I also searched around a while. Using PHP(_ADMIN)_VALUE seems to work really inconsistent and there is no official documentation about it. Also it seems to be handled different, depending on where it is located. I even found threads where it was JUST handled within location directive and not/wrong outside. Also it just works with php-fpm, not php-cgi, but with DietPi this should be the case anyway.

@Bridouz the method using nextcloud.ini worked for you? Otherwise I guess it would be okay to just set these values inside php.ini/opcache.ini/apcu.ini globally unil we found a safe working method to make this work per directory on nginx.

E.g. https://bugs.php.net/bug.php?id=51595

MichaIng commented:
I just reviewed all nginx configurations by DietPi, especially to watch out for possible other PHP_VALUE statements that would cause problems/overwriting. But I found this: https://github.com/Fourdee/DietPi/blob/testing/dietpi/conf/nginx.site-available-default#L32

So from `sites-dietpi` there seem to be only files included with `.config` endings. Might it be that `nextcloud.conf` is just ignored at the moment and thus all the PHP values don't work/just work inside main nginx conf? So renaming `nextcloud.conf` to `nextcloud.config` would solve the problem, or did I missed something? 😄 

€: ownCloud installation also uses `owncloud.config`: https://github.com/Fourdee/DietPi/blob/testing/dietpi/dietpi-software#L8422

Fourdee commented:
Hi, apologies, not been following this as much as i'd like to (time).

Confirmed, failed install:
```
E: Command line option 'i' [from -intl] is not known.
```

Appreciate the fixes @Bridouz 👍 

>  Opcache error on nextcloud admin page (Fix available, waiting for discussion about it )

Is this still pending with Nginx/All?

jbz-dev commented:
@Fourdee

Just need to recheck the whole thing on a clean install.

tweaked so many things that I broke my nginx server :)

will share the result in an hour.

jbz-dev commented:
OK I'm now unable to reach my nextcloud instance. Dont know what's wrong.

EDIT : sorry @Fourdee but I give up with Stretch/Nginx/Nextcloud ^^ Worked once and I'm not able to reproduce the correct configuration.

I will go back on Jessie tomorrow to see if I can push a working config'.

jbz-dev committed:
Fix Nginx Nextcloud config

- Disable fastcgi_request_buffering for now as it's supported since nginx 1.7 and Jessie is at 1.6.x
- Automatically load Nginx config from `sites-dietpi/nextcloud.config

jbz-dev commented:
@Fourdee 

nginx config is finally running, only opcache is pending but I don't have time for this.

MichaIng commented:
@Fourdee: Any idea why `AGI "$php_package_name"-intl` doesn't work here? `AGI ${php_package_name}-intl` would be an alternative to test, but both work for me on the following test file (copy&past from dietpi-software), as well as this is used for all the other php modules in dietpi-software during php installation:
```
#!/bin/bash

php_package_name='php'

        AGI(){

                local string="$@"

                force_options='--allow-downgrades --allow-remove-essential --allow-change-held-packages --allow-unauthenticated'

                #-qq can add a slight period of appearing nothing is happening, lets inform user
                /DietPi/dietpi/func/dietpi-notify 2 "Apt-get is processing, please wait...\n"
                DEBIAN_FRONTEND=noninteractive apt-get install -y $force_options $string

                local result=${PIPESTATUS[0]}
                if (( $result != 0 )); then

                        echo "Error"

                fi

        }

AGI ${php_package_name}-intl
```
And `./test`

@Bridouz so opcache and most likely all the php module settings do still not apply? Also this does not work by using the following minimal version of the string?
```
fastcgi_param PHP_ADMIN_VALUE "opcache.enable_cli=1
opcache.interned_strings_buffer=8
opcache.max_accelerated_files=10000
opcache.memory_consumption=128
opcache.save_comments=1
opcache.revalidate_freq=1
apc.enable_cli=1";
```
If all of this does not work, I guess I will add it to the default php configuration files for nginx (and lighttpd). They are applied globally then, but I guess it is better than shipping a Nextcloud installation with admin panel warnings.

One thing still missing: Removing `nextcloud.config` on Nextcloud uninstallation.
And does enabling the php modules now work without showing an error/missing module? Should be also tested for Jessie explicitly.

MichaIng commented on dietpi/dietpi-software:
```diff
@@ -8540,7 +8545,8 @@ _EOF_
             fastcgi_intercept_errors on;
-            fastcgi_request_buffering off;
+            # Disable because Jessie Nginx does not support that parameter
+	    #fastcgi_request_buffering off;
```
Found the related report on nextcloud forum:
https://help.nextcloud.com/t/nc-not-running-on-nginx/656/14
We should keep this in mind, once we fully switch to stretch, or we can now `sed` this based on `$DISTRO` in script?

Fourdee commented:
@MichaIng 

> Any idea why AGI "$php_package_name"-intl doesn't work here? 

Yep, its because the variable is defined as local, inside the main PHP install section (not global).  When we get to ```INSTALLING_INDEX=114```, the local variable does not exist.

I'll get this merged, then set ```php_package_name``` to a global var.

Fourdee commented:
@Bridouz 
Thanks again for the fixes 👍

MichaIng commented:
> Yep, its because the variable is defined as local

Ouuu, I am blind 😆. So no wonder why also enabling the mods didn't work. I guess it is okay to add it as global variable, so all other installations can use it which makes the script a bid smaller.
